### ISSUE TITLE
Fulfillment example code is obsolete

### ISSUE THREAD

**Issue by alexhauser** [nozzlegear/ShopifySharp#828]
The fulfillment example code in the `readme.md` does not compile any more, since it does not incorporate the breaking changes introduced with the switch to the `2022-07` API.

https://github.com/nozzlegear/ShopifySharp#fulfillments

**Comment by alexhauser** [comment]
In particular, this code in my app is not working anymore since switching to the latest release (I was a bit behind in releases, so this might as well have been caused by a breaking change in a previous update).

```C#
var fulfillment = new Fulfillment()
{
    TrackingCompany = "DPD",
    TrackingUrl = String.Format(Settings.Default.ShopifyDPDTrackingUrl, mTrackingNr),
    TrackingNumber = mTrackingNr,
    LocationId = mShopconfig.FulfillmentLocationId, 
    NotifyCustomer = true
};

fulfillment = await mFulfillmentService.CreateAsync(mOrder.Id.Value, fulfillment);
```

Can anyone please give a hint on how this code should look like nowadays?

I am aware that I can create `FulfillmentShipping` and pass that to `FulfillmentService.CreateAsync()`, however I don't seem to understand how that then relates to a particular order, since nowhere in this process I need to pass in an order id.

Thanks for the awsome work in this project btw!

**Comment by nozzlegear** [comment]
Thanks for the heads up! That definitely needs to be updated. Fulfillments are slightly more complicated now, you need to work with the `FulfillmentOrderService` and the `FulfillmentService` to fulfill things. To put it simply, once an order is created Shopify will create one or more fulfillment orders automatically for the order and then it's up to the fulfiller to pick out those fulfillment orders, add line items to them, and then use the `FulfllmentShipping` class to create a fulfillment.

[You can find our tests for the `FulfillmentService` right here](https://github.com/nozzlegear/ShopifySharp/blob/6169078e4c2452e54ee0fef08d4f727fa78fea69/ShopifySharp.Tests/Fulfillment_Tests.cs), but here's some example code for how you'd find fulfillment orders and create a fulfillment:

```cs
var fulfillments = new FulfillmentService(domain, accessToken);
var fulfillmentOrders = new FulfillmentOrderService(domain, accessToken);
var orderId = 123456;

// Find open fulfillment orders for this order
var openFulfillmentOrders = await fulfillmentOrders.ListAsync(orderId);
openFulfillmentOrders = openFulfillmentOrders.Where(f => f.Status == "open").ToList();

// Fulfill the line items
var lineItems = openFulfillmentOrders.Select(o => new LineItemsByFulfillmentOrder
{
  FulfillmentOrderId = o.Id.Value
  // Optionally specify a list of line items if you're doing a partial fulfillment
  // FulfillmentRequestOrderLineItems = ...
});
var fulfillment = await fulfillments.CreateAsync(new FulfillmentShipping
{
  Message = "items are shipping!",
  FulfillmentRequestOrderLineItems = lineItems,
  NotifyCustomer = true,
  TrackingInfo = new TrackingInfo
  {
    Company = "UPS",
    Url = "...",
    Number = "...
  }
});
```

Let me know if that answers your question!

**Comment by alexhauser** [comment]
Thank you so much, that's exactly what I was looking for!

**Comment by nozzlegear** [comment]
Great! I'll keep this open as a reminder for me to get the fulfillment example code updated.

**Comment by bQvle** [comment]
Thanks for the update, What is the practice when we have multipe TrackingNumbers/Urls? before we could send them as an array. Now the FulfillmentShipping only supports a single "TrackingInfo".

I could create multiple fulfillments for each tracking number, but this is a problem for many of my customers, because this will trigger some payment providers to create multiple invoices.

**Comment by alexhauser** [comment]
For anyone who might stumble across this issue after ugrading to the 2022-07 API-version: 

If `fulfillmentOrders.ListAsync(orderId)` comes back empty and returns no `FulfillmentOrder`s, the problem for me was that I had not granted my app access to some of the newly added fulfillment-related API-permissions like `read_merchant_managed_fulfillment_orders` or `write_merchant_managed_fulfillment_orders`.

It's weird that the call just succeeds and returns an empty array, instead of giving a permission error, but that's exactly what happens. (I know this is the Shopify API's fault, ShopifySharp has nothing to do with this). 

Hope this will save someone the hassle of figuring this out the hard way. 😉

**Comment by FlySociety** [comment]
I noticed in the fulfillment example code, there is a ToDo list which is still incomplete. For example as of writing this, 'Mark fulfillment order as open' is not checked as complete. Does that mean the code still needs working on, or just the documentation?

**Comment by alexhauser** [comment]
@nozzlegear, my app uses the current `2022-07` API release of ShopifySharp and I'm following the fulfillment example code that you have provided a few posts above this one (so NOT the deprecated one in the readme).

But today, I was notified by Shopify, that...
>TL;DR: Your custom app is leveraging the deprecated Fulfillment API. Please note that the Fulfillment API will stop working as intended on March 31st, 2023 in favor of our new Fulfillment Order API.

Is it possible that the `FulfillmentService` class or any other part of the library is still using the now deprecated `Fulfillment API` under the hood?

This is the code I'm using:

```C#
private async Task FulfillOrder()
{
	var openFulfillmentOrders = await mFulfillmentOrderService.ListAsync(mOrder.Id.Value);
	openFulfillmentOrders = openFulfillmentOrders.Where(f => f.Status == "open").ToList();

	var lineItems = openFulfillmentOrders.Select(o => new LineItemsByFulfillmentOrder()
	{
		FulfillmentOrderId = o.Id.Value
	});

	var fulfillmentShipping = new FulfillmentShipping()
	{
		NotifyCustomer = true,
		Message = "Ihre Bestellung wird nun versendet!",
		FulfillmentRequestOrderLineItems = lineItems,
		TrackingInfo = new TrackingInfo()
		{
			Company = "DPD",
			Url = String.Format(Settings.Default.ShopifyDPDTrackingUrl, mTrackingNr),
			Number = mTrackingNr
		}
	};

	var fulfillment = await mFulfillmentService.CreateAsync(fulfillmentShipping);
}
```
Thanks in advance for your input!

**Comment by nturini-cascinanet** [comment]
> Thanks for the heads up! That definitely needs to be updated. Fulfillments are slightly more complicated now, you need to work with the `FulfillmentOrderService` and the `FulfillmentService` to fulfill things. To put it simply, once an order is created Shopify will create one or more fulfillment orders automatically for the order and then it's up to the fulfiller to pick out those fulfillment orders, add line items to them, and then use the `FulfllmentShipping` class to create a fulfillment.
> 
> [You can find our tests for the `FulfillmentService` right here](https://github.com/nozzlegear/ShopifySharp/blob/6169078e4c2452e54ee0fef08d4f727fa78fea69/ShopifySharp.Tests/Fulfillment_Tests.cs), but here's some example code for how you'd find fulfillment orders and create a fulfillment:
> 
> ```cs
> var fulfillments = new FulfillmentService(domain, accessToken);
> var fulfillmentOrders = new FulfillmentOrderService(domain, accessToken);
> var orderId = 123456;
> 
> // Find open fulfillment orders for this order
> var openFulfillmentOrders = await fulfillmentOrders.ListAsync(orderId);
> openFulfillmentOrders = openFulfillmentOrders.Where(f => f.Status == "open").ToList();
> 
> // Fulfill the line items
> var lineItems = openFulfillmentOrders.Select(o => new LineItemsByFulfillmentOrder
> {
>   FulfillmentOrderId = o.Id.Value
>   // Optionally specify a list of line items if you're doing a partial fulfillment
>   // FulfillmentRequestOrderLineItems = ...
> });
> var fulfillment = await fulfillments.CreateAsync(new FulfillmentShipping
> {
>   Message = "items are shipping!",
>   FulfillmentRequestOrderLineItems = lineItems,
>   NotifyCustomer = true,
>   TrackingInfo = new TrackingInfo
>   {
>     Company = "UPS",
>     Url = "...",
>     Number = "..."
>   }
> });
> ```
> 
> Let me know if that answers your question!


Hi @nozzlegear , according to you _"once an order is created Shopify will create one or more fulfillment orders automatically for the order"_ if i call "ListAsync" function from FulfillmentOrderService i must receive some response, but my response list is empty; maybe it's because i don't use 2022-07 api? where i can see the API version i'm using?

**Comment by patrickebates** [comment]
Am I missing how/where to set the Fulfillment Location with the new objects?

**Comment by brujah** [comment]
> Am I missing how/where to set the Fulfillment Location with the new objects?

same here

**Comment by ajvanlaningham** [comment]
A NOTE ABOUT PARTIAL FULFILLMENT: 
It took a day for me to figure out so I wanted to expand on this thread for anyone who might find it helpful.

`var fulfillments = new FulfillmentService(domain, accessToken);
var fulfillmentOrders = new FulfillmentOrderService(domain, accessToken);
var orderId = 123456;
//*** NEW CODE ***
var lineItemId = 654321; //the long? ShopifySharp.LineItemId of the item that you want to partially fulfill

// Find open fulfillment orders for this order
var openFulfillmentOrders = await fulfillmentOrders.ListAsync(orderId);
openFulfillmentOrders = openFulfillmentOrders.Where(f => f.Status == "open").ToList();

//grab the open fulfullment order
var fulfullmentLineItems = openFulfillments.FirstOrDefault().FulfillmentOrderLineItems.ToList();
var fulfillmentLineItem =  fulfillmentLineItems.Find(li => li.lineItemId == lineItem.Id); //pick the correct line item

// Fulfill the line items
var lineItems = openFulfillmentOrders.Select(o => new LineItemsByFulfillmentOrder
{
  FulfillmentOrderId = o.Id.Value
  //***NEW CODE***
  //Specify a list of line items if you're doing a partial fulfillment for
  FulfillmentRequestOrderLineItems = new List<FulfullmentRequestOrderLineItems>
   {
         new ShopifySharp.FulfillmentRequestOrderLineItem
          {
             Id = fulfillmentLineItem.Id,  //NOTE: this is not fulfillmentLineItem.LineItemId!! 
             Quantity = ... // The quantity you want to partially fulfill.
          }
    };
});
var fulfillment = await fulfillments.CreateAsync(new FulfillmentShipping
{
  Message = "items are shipping!",
  FulfillmentRequestOrderLineItems = lineItems,
  NotifyCustomer = true,
  TrackingInfo = new TrackingInfo
  {
    Company = "UPS",
    Url = "...",
    Number = "..."
  }
});`

It took me a fair amount of time to figure out why I was getting an error, and ultimately it was because I was using the "FulfillmentOrderLineItem.LineItemId" instead of the "FulfillmentOrderLineItem.Id".

I can't see anything wrong with using the above code always, especially if you are doing work for a shopify store front in which partial fulfillments are a regular occurance. If it's a full fulfillment, just loop through all the items  in the fulfillmentLineItems list and fill the full quantity. I'd love to hear from a more experienced developer if I'm wrong though!

**Comment by nozzlegear** [comment]
Thanks Alex! I actually ran into that issue myself recently, I'll update the readme soon with a note about the line item ids specifically.

--
Joshua Harms

**Comment by z0ha1b** [comment]
I am having 404 error while executing the following code.
 var lineItems = fulfillment.LineItems.Select(x => new FulfillmentRequestOrderLineItem
            {
                Id = x.Id,
                Quantity = x.Quantity,
            });

            var service = new FulfillmentService(acct.ShopifyURL, acct.ShopifyAccessToken);
            var fulfillmentShipping = new FulfillmentShipping()
            {
                Message = "Items will be shipped now.",
                NotifyCustomer = fulfillment.NotifyCustomer,
                TrackingInfo = new TrackingInfo
                {
                    Company = fulfillment.TrackingCompany,
                    Url = fulfillment.TrackingUrl,
                    Number = fulfillment.TrackingNumber,
                },
                FulfillmentRequestOrderLineItems = new List<LineItemsByFulfillmentOrder>()
                {
                    new LineItemsByFulfillmentOrder()
                    {
                        FulfillmentOrderId = orderId,
                        FulfillmentRequestOrderLineItems = lineItems.ToArray()
                    }
                }
            };

            fulfillment = await service.CreateAsync(fulfillmentShipping);

**Comment by nozzlegear** [comment]
@z0ha1b That looks correct I think. One thing I'd investigate is that `orderId` you have -- are you positive that's the fulfillment order's id, and not the full order's id? It has to be the fulfillment order.

**Comment by z0ha1b** [comment]
> @z0ha1b That looks correct I think. One thing I'd investigate is that `orderId` you have -- are you positive that's the fulfillment order's id, and not the full order's id? It has to be the fulfillment order.

thanks, it worked fine :)

**Comment by nozzlegear** [comment]
Hey @PendelinP! Sorry to hear you're having trouble. If the code from the tests didn't work, I think my first suspicion would be to check the shop domain that you're using. Shopify will redirect requests and potentially cause 404s if you're not using the right myshopify domain. 

Make sure the domain you're passing in to the service constructors looks like `example.myshopify.com` and not `example.com`. On top of that, we've recently found that even the myshopify domain can get redirected if a store has more than one of them. Check issue #968 and make sure that's not the case for you!

**Comment by PendelinP** [comment]
> Hey @PendelinP! Sorry to hear you're having trouble. If the code from the tests didn't work, I think my first suspicion would be to check the shop domain that you're using. Shopify will redirect requests and potentially cause 404s if you're not using the right myshopify domain.
> 
> Make sure the domain you're passing in to the service constructors looks like `example.myshopify.com` and not `example.com`. On top of that, we've recently found that even the myshopify domain can get redirected if a store has more than one of them. Check issue #968 and make sure that's not the case for you!

Wow - that was super fast and this was exactly my problem. I'm soooooo thankful! I would have never checked this :D Again, thank you so, so much!

**Comment by nozzlegear** [comment]
Awesome, glad it's working now @PendelinP!

**Comment by lcost0323** [comment]
Hello, would anyone know why all my fulfillment orders return as closed?

**Comment by nozzlegear** [comment]
Hey @lcost0323, I think I've run into this before. If I recall correctly, there's a setting that new Shopify stores have where Shopify will automatically fulfill all fulfillment orders as soon as an order is placed. The store owner can turn this setting off in the store by going to the store's **Settings Page -> **Checkout** -> **Order Processing** and selecting the option that says "**Don't fulfill any of the order's line items automatically**".

<img width="545" alt="2024-06-04T12-51-40-05-00" src="https://github.com/nozzlegear/ShopifySharp/assets/2417276/4d8200fb-a8e9-4227-b3d0-27f0e82c6700">

<img width="757" alt="2024-06-04T12-52-36-05-00" src="https://github.com/nozzlegear/ShopifySharp/assets/2417276/2cead355-aab2-4dfd-b99e-d15603f619c5">

Alternatively, if you don't control the store (i.e. you're building an app for the Shopify App Store), you'll need to create a Fulfillment Service using the Fulfillment Service API instead. With the fulfillment service you can assign the store's products/inventory to your new fulfillment service's location, and after that Shopify should no longer fulfill them for you. That's kind of a simple overview of a fulfillment service, there's a lot more setup to it than that, but that's the gist of it.

Let me know if that solves the problem for you or if you have any questions about fulfillment services. Happy to help!

**Comment by lcost0323** [comment]
Hi @nozzlegear , thank you for getting back to me so quickly. It looks like we have it set to the gift card option and the order itself says unfulfilled. When these closed fulfillment orders come over, they also say unsubmitted.   What I really want to do is fulfill, and then mark the order shipped/add tracking information so it gets closed out, but it seems like I have to create some kind of fulfilment order prior to updating the tracking information. I am very new to shopify.

**Comment by lcost0323** [comment]
This turned out to be a "me" issue.. I was using the wrong orderID, it is working now.

**Comment by nozzlegear** [comment]
@lcost0323 Awesome, glad you got it working! Sorry I didn't reply to your follow up yesterday, I set a reminder to do so this morning but it looks like it's not necessary! Let me know if you have any other questions or problems getting it working. You can email me at the email address in my GitHub profile.

**Comment by jhunexjun** [comment]
@nozzlegear, I am still using ShopifySharp v5.16.0.0 which is according to table, uses Shopify API version 2022-04.
Using that version, will the code still work below? Or Shopify already eliminated the 2022-04?
Because I am getting error **_Shopify returned 406 Not Acceptable but ShopifySharp was unable to parse the response JSON_** in the _fulfillmentSvc.CreateAsync()

```
            var toFulfill = fullOrder[0].LineItems
                .Where(li => items.Contains(li.Id.Value))
                // .Where(li => items.Contains(li.VariantId.Value))    // use the variant Id instead of Id value.
                .Where(li => fullOrder[0].Fulfillments.Where(f => f.Status == "pending")
                .Any(f => f.LineItems.Any(fLineItem => fLineItem.Id == li.Id)) == false);
            // .Any(f => f.LineItems.Any(fLineItem => fLineItem.VariantId == li.VariantId)) == false);


            //Create a new fulfillment that will fulfill the remaining line items
            var notifyCustomer = false;
            var trackingNumber = Guid.NewGuid().ToString();
            var fulfillment = new Fulfillment()
            {
                TrackingCompany = "AuntieDot Shipping, LLC.",
                TrackingNumber = trackingNumber,
                TrackingUrl = "https://shipping-company.com/track/" + trackingNumber,
                OrderId = fullOrder[0].Id.Value,
                LineItems = toFulfill,
                NotifyCustomer = notifyCustomer
            };

            //Send the fulfillment to Shopify            
            fulfillment = await _fulfillmentSvc.CreateAsync(fullOrder[0].Id.Value, fulfillment);
            if (fulfillment.Status != "pending" && fulfillment.Status != "success")
            {
                // Status = 'cancelled': The fulfillment was cancelled.
                // Status = 'error': There was an error with the fulfillment. (No details provided!)
                // Status = 'failure': The fulfillment request failed. (No details provided!)
                throw new Exception("Failed to fulfill line items.");
            }
```
Do I need to update my version?

**Comment by nozzlegear** [comment]
Hey @jhunexjun! Yes, you'd need to update ShopifySharp. The API version 2022-04 has been taken offline, so most API calls using that version will fail.

**Comment by joshbybee** [comment]
@jhunexjun you're right, I updated my comment for the changes I had to make when upgrading the API. You can refer to my example above if you'd like: https://github.com/nozzlegear/ShopifySharp/issues/828#issuecomment-1593706002

**Comment by jhunexjun** [comment]
@nozzlegear I am having a hard time with the Id you're referring to. My setup is I do not download the FulfillmentOrderLineItem to get the `Id` and choose them through User Interface but I have a third-party application that through their `variant Id` I match them.
Do these code works?
```
async Task ProcessShopifyOrder(PsDocHeader psDocHdr)
{
    List<PsDocHeaderLines> parentLineItems = await _counterpoint.GetFulfilledParentLineItems(psDocHdr.CP_ORDER_DOC_NO);
    var fulfillmentOrders = await _fulfillmentOrderSvc.ListAsync(psDocHdr.SHOPIFY_ORDER_ID.Value);
    fulfillmentOrders = fulfillmentOrders.Where(f => f.Status == "open");

    foreach(FulfillmentOrder fulfillmentOrder in fulfillmentOrders)
    {
        var itemIdsAa = (await _fulfillmentSvc.ListAsync(psDocHdr.SHOPIFY_ORDER_ID.Value)).Items;
        var s = itemIdsAa.SelectMany(o => o.LineItems);

        var fulfillmentOrderLineItems = new LineItemsByFulfillmentOrder
        {
            FulfillmentOrderId = fulfillmentOrder.Id,
            FulfillmentRequestOrderLineItems = s.Select(itemId => new FulfillmentRequestOrderLineItem
            {
                Id = itemId.Id,
                Quantity = GeFulfilledtQuantity(s, parentLineItems),
            }),
        };


        var fulfillment = await _fulfillmentSvc.CreateAsync(new FulfillmentShipping
        {
            Message = "Items are fulfilled by Counterpoint.",
            FulfillmentRequestOrderLineItems = new[] { fulfillmentOrderLineItems },
            NotifyCustomer = false,  // move this to a system variable.
            //TrackingInfo = new TrackingInfo // can be eliminated.
            //{
            //    Company = "UPS",
            //    Url = "https://www.ups.com",
            //    Number = Guid.NewGuid().ToString()
            //}
        });

        if (fulfillment.Status == "success")
        {
           // do something
        }
    } 
}

long GeFulfilledtQuantity(IEnumerable<LineItem> lineItems, List<PsDocHeaderLines> cpLineItems)
{
    foreach(LineItem lineItem in lineItems)
        for(int y = 0; y < cpLineItems.Count; y++)
            if (lineItem.VariantId == cpLineItems[y].VariantId)
                return cpLineItems[y].QtyInvoiced;

    return 0;
}
```
I am having a hard time with this code above:
```
// Where request = some kind of form post or json model
long[] itemIds = request.ItemsToFulfill;
```
because there's no sample.
Thank you and more powers.
